### PR TITLE
Improve error handling on GROBID server connection issues

### PR DESCRIPTION
**Pull request by nitram509**
Hi,

This PR relates to issues #6517 and #6891.
It's more improvement than an actual fix because the root cause is: the server is not available (which I can't fix).

Motivation:
When I read #6517 and #6891, I realized, that these were reported by 'power users',
means people who e.g. understand how to activate debug logging. 
When I tried to reproduce the issues, I realized, that for the end-users the issue is not transparent. 
Additionally, on my OSX, the Java 14 runtime has an infinite connect timeout set, so I never got a visual response on my action (refers to "New entry from plain text" button)

What I changed:
* add a property connectTimeout to URLDownloader
   * this avoids the situation, where e.g. on similar setups than mine (Java 14.0.2 on OSX), there's an infinite default connectTimeout and I never saw any error in UI nor log.
   * set the default connectTimeout to 30 seconds. Which is e.g. the same order of magnitude as in many other tools and thus should have no side effects to other usages of this class
   * set the connectTimeout for GROBID server to 5 seconds, because this should be sufficient as I expect this is a cloud-based service - please comment, if this assumption is wrong.
* refactored the GrobidServce so that IOExceptions are percolating higher in the call-stack and then can be handled in the UI
* in case of error, I add a notification to the user, with an explicit message, so users have a chance to know, what's going on.. IMHO, that's more correct and easier to understand for users, over the current behavior, where just a message "0 entries parsed" occurs. (screenshot below in the comments)
* added EN and DE translation for the new message.

I hope, with this new behavior, users will better understand, why this feature does not work as expected. 

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

Any feedback welcome.

### CODE REVIEW

**Comment by nitram509** [comment]
Here's a screenshot, how the error message looks like (the red arrow is just for demonstration purpose).

File: src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java:69]: I think it's not so important for the user which exact url failed (especially since the popup disappears to quickly to type it in the url of a webbrower to check). So I suggest the following wording: "Could not connect to JabRef server (Connection timed out)."

File: src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java:43]: Please use `ParserException` instead of the uncheckedioex.

File: src/main/java/org/jabref/logic/net/URLDownload.java
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/net/URLDownload.java:322]: what about `connectionTimeout.toMillis()`?

File: src/test/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcherTest.java
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcherTest.java:119]: Very good that you added a test as well. Small suggestion for improvement: use `Assertions.assertThrows` to ensure that the exception is indeed thrown.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Thanks a lot for your PR. The code looks already very good. I've a few minor remarks. Could you please address them?

File: src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java:69]: And maybe add the details via error log.

Comment (nitram509) [reply src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java:69]: Fair point, @tobiasdiez.
This should be "Could not connect to GROBID server (Connection timed out)." Shouldn't it?

File: src/test/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcherTest.java
Comment (nitram509) [reply src/test/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcherTest.java:119]: Will have a look. Thank you.

File: src/main/java/org/jabref/logic/net/URLDownload.java
Comment (nitram509) [reply src/main/java/org/jabref/logic/net/URLDownload.java:322]: Good catch. Quite convenient and overseen, thank you.

File: src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java
Comment (nitram509) [reply src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java:43]: Hmm 🤔 
But that's not a ParserException, we are not even at the point of parsing yet.
Also, ParserException is a checked one, which has way more side-effects in terms of refactoring than an unchecked one.
Since you're one of the maintainers ... may I ask for your advise?

Here's the situation: the fetcher implementation before my PR did catch all exceptions (especially all IO related ones).
When an IOException is bubbling up in the call hierarchy.
The SearchBasedFetcher.performSearch method declaration does not mention any exception.
So when I want to throw e.g. a ParserException (I would prefer IOException because of its technical nature, though)
then I need to adapt the interface/abstract class and this will lead to many more files changed.

Would this be in your interest and aligned with your general architecture, @tobiasdiez ?

File: src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java:69]: I would still say "JabRef server"...the user knows what Jabref is, but not necessarily what grobid is. I would see this also as an "implementation detail", maybe in the future we switch the backend from grobid to something else..

File: src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java
Comment (tobiasdiez) [reply src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java:43]: Good catch, not sure where I was with my thoughts... I meant of course the `FetcherException` instead of `ParserException`. The general idea is to use `FetcherException` for everything that is semi-forseeable (e.g. connection time outs, bad response errors etc). So that would be the case here as well. 

You are right that it might lead to a bit more refactoring as it's a checked exception, but I see this as an advantage instead of an disadvantage. 

Note also that you should provide the cause of the `FetcherException` already when it's created (since that's usual the point where you can figure out what went wrong).

Comment (nitram509) [reply src/main/java/org/jabref/logic/importer/fetcher/GrobidCitationFetcher.java]: I did follow the idea and used the FetcherExeption. That was simple.
There's one caveat though, because of Java Stream don't play well with checked exceptions, I had to wrap IOException in RuntimeException. Since this is wrapped and un-wrapped within the same class, I hope you agree, in this is acceptable.
The exception thrown by the fetcher is a FetcherException, as discussed.

File: src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java
Comment (nitram509) [reply src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java]: It was tricky to get the Message to a phrase as you proposed.
I reduced the level of detail towards:
"There are connection issues with a JabRef server. Detailed information: %0."
(See latest screenshot)

Is this OK for you @tobiasdiez ?

**Comment by nitram509** [comment]
All reviews incorporated.
May I ask for another review, please?

/cc @tobiasdiez

File: src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/bibtexextractor/BibtexExtractorViewModel.java]: Perfect! Thanks

**Review by tobiasdiez — APPROVED** [review]
Thanks a lot for the quick follow-up! From my side, this is now good to go. We have the convention that at least two core developers have to approve a PR, so please have a bit of patience until the second review arrives.

@koppor The author check fails again...should we disable it again?

**Comment by nitram509** [comment]
> Thanks a lot for the quick follow-up! From my side, this is now good to go. We have the convention that at least two core developers have to approve a PR, so please have a bit of patience until the second review arrives.
> 
> @koppor The author check fails again...should we disable it again?

Thank you very much, @tobiasdiez 
I'm participating in the Hacktoberfest.
May I ask, to label this PR as "hacktoberfest-accepted", to signal that I did follow the rules, please?

**Comment by Siedlerchr** [comment]
The PR wil count towards hacktoberfest because we applied the hacktoberfest label on the repo.

**Review by Siedlerchr — APPROVED** [review]

**Comment by Siedlerchr** [comment]
Thanks for your contribution! Looking foward seeing more from you ;)

**Pull request merged**
